### PR TITLE
fix(CORE/Commands): .cheat god lowering hp to 1 and .cheat power not refilling power

### PR DESCRIPTION
**Pull request by peti446**
## CHANGES PROPOSED:
-  On power resource update to have maximum value if .cheat power is enabled
-  Dont check if player has enough power to cast spell if .cheat power is enabled
-  Changed SendSpellNonMelee...Log to send damage as absorb instead of damage to avoid client prediction lowering hp bar when .cheat god is enabled but still be able to debug damage from enemies (looking at absorb log)
- Avoid fall damage while .god mode is enabled

## ISSUES ADDRESSED:

- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/3257
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/1534

## TESTS PERFORMED:
- Activated god mode (.cheat god) and infinite power (.cheat power) on a warrior, dk, rouge and mage
- Went to ICC Raid
- Pulled all till first boss and all attacking for 5 minutes
- Checked my hp bar is 100%
- Attacked mobs with many different spells that consume power
- Checked that power bar ddi not get below 95%

## HOW TO TEST THE CHANGES:

For .cheat god
1. Activate god mode (.cheat god)
2. Go to ICC or any other instance zone with many mobs
3. Pull many mobs and let them attack for  a while
4. Check that HP bar did not go down

For .cheat power
1. Create class (warrior, DK, rouge and any Mana clasS)
2. Activate .cheat power on each
3. Use spells
4. Check that resource did not go down and its always at 95%-100% even outside combat.

## KNOWN ISSUES AND TODO LIST:

- [ ]
- [ ] 

## Target branch(es):
- [x] Master

 
## How to test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

File: src/server/game/Entities/Unit/Unit.cpp [hunk 1]
```diff
@@ -5636,14 +5654,26 @@ void Unit::SendPeriodicAuraLog(SpellPeriodicAuraLogInfo* pInfo)
...
+                absorb = damage;
+                damage = 0;
+            }
+
+
+            data << uint32(damage);                  // damage
             data << uint32(pInfo->overDamage);              // overkill?
             data << uint32(aura->GetSpellInfo()->GetSchoolMask());
-            data << uint32(pInfo->absorb);                  // absorb
+            data << uint32(absorb);                  // absorb
```
Comment (Kitzunu) [inline src/server/game/Entities/Unit/Unit.cpp:5673]: Fix space for comment

File: src/server/game/Entities/Unit/Unit.cpp [hunk 2]
```diff
@@ -5710,28 +5740,37 @@ void Unit::SendAttackStateUpdate(CalcDamageInfo* damageInfo)
     sLog->outDebug(LOG_FILTER_UNITS, "WORLD: Sending SMSG_ATTACKERSTATEUPDATE");
 #endif
 
+    //IF we are in cheat mode we swap absorb with damage and set daamge to 0, this way we can still debug damage but our hp bar will not drop
```
Comment (Kitzunu) [inline src/server/game/Entities/Unit/Unit.cpp:5743]: daamage -> damage

File: src/server/game/Entities/Unit/Unit.cpp [hunk 3]
```diff
@@ -5636,14 +5654,26 @@ void Unit::SendPeriodicAuraLog(SpellPeriodicAuraLogInfo* pInfo)
     data << uint32(aura->GetAuraType());                    // auraId
     switch (aura->GetAuraType())
     {
-        case SPELL_AURA_PERIODIC_DAMAGE:
-        case SPELL_AURA_PERIODIC_DAMAGE_PERCENT:
-            data << uint32(pInfo->damage);                  // damage
+    case SPELL_AURA_PERIODIC_DAMAGE:
+    case SPELL_AURA_PERIODIC_DAMAGE_PERCENT:
```
Comment (Kitzunu) [inline src/server/game/Entities/Unit/Unit.cpp:5658]: indentation

File: src/server/game/Entities/Unit/Unit.cpp [hunk 4]
```diff
@@ -5561,15 +5562,24 @@ void Unit::SendSpellNonMeleeReflectLog(SpellNonMeleeDamage* log, Unit* attacker)
...
     data.append(attacker->GetPackGUID());
     data << uint32(log->SpellID);
-    data << uint32(log->damage);                            // damage amount
-    int32 overkill = log->damage - log->target->GetHealth();
+    data << uint32(damage);                            // damage amount
+    int32 overkill = damage - log->target->GetHealth();
     data << uint32(overkill > 0 ? overkill : 0);            // overkill
     data << uint8 (log->schoolMask);                        // damage school
-    data << uint32(log->absorb);                            // AbsorbedDamage
+    data << uint32(absorb);                            // AbsorbedDamage
```
Comment (Kitzunu) [inline src/server/game/Entities/Unit/Unit.cpp:5582]: comment spacinbg

File: src/server/game/Entities/Player/Player.cpp [hunk 5]
```diff
@@ -2666,6 +2666,21 @@ void Player::Regenerate(Powers power)
...
         return;
 
+    //If .cheat power is on alwas have the max power
+    if (GetCommandStatus(CHEAT_POWER))
+    {
+        if (m_regenTimerCount >= 2000)
+        {
+            //Set the value to 0 first then set it to max to force resend of packet as for range clients keeps removing rage
+            if(power == POWER_RAGE || power == POWER_RUNIC_POWER)
+                UpdateUInt32Value(UNIT_FIELD_POWER1 + power, 0);
```
Comment (FrancescoBorzi) [inline src/server/game/Entities/Player/Player.cpp:2676]: let's add brackets here, it's safer:

```
if(power == POWER_RAGE || power == POWER_RUNIC_POWER)
{
                UpdateUInt32Value(UNIT_FIELD_POWER1 + power, 0);
}
```

**Comment by Si1ker** [comment]
@peti446 Would you be able to resolve the conflicts in this PR? Apologize we didn't catch it at the time with the importance it deserves.

**Comment by peti446** [comment]
@Si1ker Done :)

**Review by Si1ker — APPROVED** [review]

**Comment by Si1ker** [comment]
Thank you very much @peti446 !
Really appreciated, we're gonna keep an eye on this 👀

**Comment by Delcidx** [comment]
TESTED, working great.
- HP bar keeps in 100% with cheat god on.
test2

- Energy/Mana keeps in 100% while using spells with cheat power on.
tested

- Rage/Runic Power does not drop below 98% with cheat power on.
tested2

File: src/server/game/Entities/Player/Player.cpp [hunk 6]
```diff
@@ -2726,6 +2726,23 @@ void Player::Regenerate(Powers power)
     if (!maxValue)
         return;
 
+    //If .cheat power is on alwas have the max power
```
Comment (Kitzunu) [inline src/server/game/Entities/Player/Player.cpp:2729]: always

**Comment by Si1ker** [comment]
@peti446 Thanks for the PR, it's a valuable contribution to the project! ⭐ 

@Delcidx Great testing report, we appreciated it very much! 🎉

File: src/server/game/Entities/Unit/Unit.cpp [hunk 7]
```diff
@@ -733,14 +733,15 @@ uint32 Unit::DealDamage(Unit* attacker, Unit* victim, uint32 damage, CleanDamage
...
-                return 0;
-
         if (pet && pet->IsAlive())
             pet->AI()->OwnerAttackedBy(attacker);
     }
 
+    //Dont deal damage to unit if .cheat god is enable.
+    if (victim->GetTypeId() == TYPEID_PLAYER)
+        if (victim->ToPlayer()->GetCommandStatus(CHEAT_GOD))
+            return 0;
```
Comment (Kitzunu) [inline src/server/game/Entities/Unit/Unit.cpp:743]: brackets

File: src/server/game/Entities/Unit/Unit.cpp [hunk 8]
```diff
@@ -5749,14 +5750,22 @@ void Unit::SendSpellNonMeleeReflectLog(SpellNonMeleeDamage* log, Unit* attacker)
...
+    {
+        absorb = damage;
+        damage = 0;
+    }
     data.append(log->target->GetPackGUID());
     data.append(attacker->GetPackGUID());
     data << uint32(log->SpellID);
-    data << uint32(log->damage);                            // damage amount
-    int32 overkill = log->damage - log->target->GetHealth();
+    data << uint32(damage);                                 // damage amount
```
Comment (Kitzunu) [inline src/server/game/Entities/Unit/Unit.cpp:5764]: blanks

File: src/server/game/Entities/Unit/Unit.cpp [hunk 9]
```diff
@@ -5749,14 +5750,22 @@ void Unit::SendSpellNonMeleeReflectLog(SpellNonMeleeDamage* log, Unit* attacker)
...
+    int32 overkill = damage - log->target->GetHealth();
     data << uint32(overkill > 0 ? overkill : 0);            // overkill
     data << uint8 (log->schoolMask);                        // damage school
-    data << uint32(log->absorb);                            // AbsorbedDamage
+    data << uint32(absorb);                                 // AbsorbedDamage
```
Comment (Kitzunu) [inline src/server/game/Entities/Unit/Unit.cpp:5768]: blanks

File: src/server/game/Entities/Unit/Unit.cpp [hunk 10]
```diff
@@ -5769,14 +5778,22 @@ void Unit::SendSpellNonMeleeReflectLog(SpellNonMeleeDamage* log, Unit* attacker)
...
+    {
+        absorb = damage;
+        damage = 0;
+    }
     data.append(log->target->GetPackGUID());
     data.append(log->attacker->GetPackGUID());
     data << uint32(log->SpellID);
-    data << uint32(log->damage);                            // damage amount
-    int32 overkill = log->damage - log->target->GetHealth();
+    data << uint32(damage);                                 // damage amount
```
Comment (Kitzunu) [inline src/server/game/Entities/Unit/Unit.cpp:5792]: blanks

File: src/server/game/Entities/Unit/Unit.cpp [hunk 11]
```diff
@@ -5769,14 +5778,22 @@ void Unit::SendSpellNonMeleeReflectLog(SpellNonMeleeDamage* log, Unit* attacker)
...
+    int32 overkill = damage - log->target->GetHealth();
     data << uint32(overkill > 0 ? overkill : 0);            // overkill
     data << uint8 (log->schoolMask);                        // damage school
-    data << uint32(log->absorb);                            // AbsorbedDamage
+    data << uint32(absorb);                                 // AbsorbedDamage
```
Comment (Kitzunu) [inline src/server/game/Entities/Unit/Unit.cpp:5796]: blanks

File: src/server/game/Entities/Unit/Unit.cpp [hunk 12]
```diff
@@ -5825,12 +5841,24 @@ void Unit::SendPeriodicAuraLog(SpellPeriodicAuraLogInfo* pInfo)
...
+                absorb = damage;
+                damage = 0;
+            }
+
+
+            data << uint32(damage);                         // damage
             data << uint32(pInfo->overDamage);              // overkill?
             data << uint32(aura->GetSpellInfo()->GetSchoolMask());
-            data << uint32(pInfo->absorb);                  // absorb
+            data << uint32(absorb);                         // absorb
```
Comment (Kitzunu) [inline src/server/game/Entities/Unit/Unit.cpp:5858]: blanks

File: src/server/game/Entities/Unit/Unit.cpp [hunk 13]
```diff
@@ -5825,12 +5841,24 @@ void Unit::SendPeriodicAuraLog(SpellPeriodicAuraLogInfo* pInfo)
...
-            data << uint32(pInfo->damage);                  // damage
+            {
+            //IF we are in cheat mode we swap absorb with damage and set damage to 0, this way we can still debug damage but our hp bar will not drop
+            uint32 damage = pInfo->damage;
+            uint32 absorb = pInfo->absorb;
+            if (GetTypeId() == TYPEID_PLAYER && ToPlayer()->GetCommandStatus(CHEAT_GOD))
+            {
+                absorb = damage;
+                damage = 0;
+            }
```
Comment (Kitzunu) [inline src/server/game/Entities/Unit/Unit.cpp:5852]: indentation of the whole thing

File: src/server/game/Spells/Spell.cpp [hunk 14]
```diff
@@ -4892,6 +4892,9 @@ SpellCastResult Spell::CheckRuneCost(uint32 RuneCostID)
         return SPELL_CAST_OK;
 
     Player* player = m_caster->ToPlayer();
+    //If we are in .cheat power mode we dont need to check the cost as we are expected to be able to use it anyways (infinite power)
+    if (player->GetCommandStatus(CHEAT_POWER))
+        return SPELL_CAST_OK;
```
Comment (Kitzunu) [inline src/server/game/Spells/Spell.cpp:4897]: brackets

File: src/server/game/Spells/Spell.cpp [hunk 15]
```diff
@@ -6471,6 +6474,12 @@ SpellCastResult Spell::CheckPower()
     if (m_CastItem)
         return SPELL_CAST_OK;
 
+    //While .cheat power is enabled dont check if we need power to cast the spell
+    if (m_caster->GetTypeId() == TYPEID_PLAYER)
+        if (m_caster->ToPlayer()->GetCommandStatus(CHEAT_POWER))
+            return SPELL_CAST_OK;
```
Comment (Kitzunu) [inline src/server/game/Spells/Spell.cpp:6480]: brackets

**Review by Kitzunu — CHANGES_REQUESTED** [review]

File: src/server/game/Entities/Unit/Unit.cpp (hunk 8 above)
Comment (Kitzunu) [reply src/server/game/Entities/Unit/Unit.cpp:5764]: my bad, I missed the comment :P

**Comment by FrancescoBorzi** [comment]
thanks @Delcidx for testing this!

**Comment by peti446** [comment]
@Kitzunu Fixed the brackets and indentation :)

**Comment by FrancescoBorzi** [comment]
good job @peti446 , thanks @Si1ker @Kitzunu for reviewing this and @Delcidx for testing. Nice team work!

**Pull request merged**